File: pkg/signaling/srs.go
package signaling

import (
	"context"

	"github.com/ossrs/go-oryx-lib/errors"
)

type Srs struct {
	Ctx  context.Context
	Addr string // The address of SRS, eg: http://localhost:1985
}

func (s *Srs) Publish(id, ssrc string) (int, error) {
	req := struct {
		Id   string `json:"id"`
		SSRC string `json:"ssrc"`
	}{
		id, ssrc,
	}

	res := struct {
		Code int `json:"code"`
		Port int `json:"port"`
	}{}

	if err := apiRequest(s.Ctx, s.Addr+"/gb/v1/publish/", req, &res); err != nil {
		return 0, errors.Wrapf(err, "gb/v1/publish")
	}

	return res.Port, nil
}

func (s *Srs) Unpublish(id string) error {
	return nil
}

//	{
//	    "code": 0,
//	    "server": "vid-y19n6nm",
//	    "service": "382k456r",
//	    "pid": "9495",
//	    "streams": [{
//	        "id": "vid-9y0ozy0",
//	        "name": "0551954854",
//	        "vhost": "vid-v2ws53u",
//	        "app": "live",
//	        "tcUrl": "webrtc://127.0.0.1:1985/live",
//	        "url": "/live/0551954854",
//	        "live_ms": 1720428680003,
//	        "clients": 1,
//	        "frames": 8431,
//	        "send_bytes": 66463941,
//	        "recv_bytes": 89323998,
//	        "kbps": {
//	            "recv_30s": 0,
//	            "send_30s": 0
//	        },
//	        "publish": {
//	            "active": false,
//	            "cid": "b3op069g"
//	        },
//	        "video": null,
//	        "audio": null
//	    }]
//	}
func (s *Srs) GetStreamStatus(id string) (bool, error) {
	type Stream struct {
		Id      string `json:"id"`
		Name    string `json:"name"`
		Publish struct {
			Active bool   `json:"active"`
			Cid    string `json:"cid"`
		} `json:"publish"`
	}
	res := struct {
		Code    int      `json:"code"`
		Streams []Stream `json:"streams"`
	}{}

	if err := apiRequest(s.Ctx, s.Addr+"/api/v1/streams?count=99", nil, &res); err != nil {
		return false, errors.Wrapf(err, "api/v1/stream")
	}

	if len(res.Streams) == 0 {
		return false, nil
	} else {
		for _, v := range res.Streams {
			if v.Name == id {
				return v.Publish.Active, nil
			}
		}
	}
	return false, nil
}

File: web/NextGB/src/api/mediaserver/srs.ts
import type { ClientInfo, StreamInfo, VersionInfo } from './types';
import { MediaServerType } from './types';
import { BaseMediaServer } from './base';
import axios from 'axios';

interface SRSVersionResponse {
    code: number;
    server: string;
    service: string;
    pid: string;
    data: {
        major: number;
        minor: number;
        revision: number;
        version: string;
    };
}

interface SRSClientsResponse {
    code: number;
    server: string;
    service: string;
    pid: string;
    clients: {
        id: string;
        vhost: string;
        stream: string;
        ip: string;
        pageUrl: string;
        swfUrl: string;
        tcUrl: string;
        url: string;
        name: string;
        type: string;
        publish: boolean;
        alive: number;
        send_bytes: number;
        recv_bytes: number;
        kbps: {
            recv_30s: number;
            send_30s: number;
        };
    }[];
}

interface SRSStreamResponse {
    code: number;
    server: string;
    service: string;
    pid: string;
    streams: {
        id: string;
        name: string;
        vhost: string;
        app: string;
        tcUrl: string;
        url: string;
        live_ms: number;
        clients: number;
        frames: number;
        send_bytes: number;
        recv_bytes: number;
        kbps: {
            recv_30s: number;
            send_30s: number;
        };
        publish: {
            active: boolean;
            cid: string;
        };
        video?: {
            codec: string;
            profile: string;
            level: string;
            width: number;
            height: number;
        };
        audio?: {
            codec: string;
            sample_rate: number;
            channel: number;
            profile: string;
        };
    }[];
}

export class SRSServer extends BaseMediaServer {
    private baseUrl: string;

    constructor(host: string, port: number) {
        super(MediaServerType.SRS);
        this.baseUrl = `http://${host}:${port}`;
    }

    async getVersion(): Promise<VersionInfo> {
        try {
            const response = await axios.get<SRSVersionResponse>(`${this.baseUrl}/api/v1/versions`);
            
            return {
                version: response.data.data.version,
                buildDate: undefined, // SRS API 没有提供构建日期
                platform: `SRS Server: ${response.data.server}` // 使用 server 标识作为平台信息
            };
        } catch (error) {
            throw new Error(`Failed to get SRS version: ${error}`);
        }
    }

    async getStreamInfo(): Promise<StreamInfo[]> {
        try {
            const response = await axios.get<SRSStreamResponse>(`${this.baseUrl}/api/v1/streams/`);
            
            return response.data.streams.map(stream => ({
                id: stream.id,
                name: stream.name,
                vhost: stream.vhost,
                url: stream.tcUrl,
                clients: stream.clients - 1,
                active: stream.publish.active,
                send_bytes: stream.send_bytes,
                recv_bytes: stream.recv_bytes,
                video: stream.video ? {
                    codec: stream.video.codec,
                    width: stream.video.width,
                    height: stream.video.height,
                    fps: 0,  // SRS API 没有直接提供 fps 信息
                } : undefined,
                audio: stream.audio ? {
                    codec: stream.audio.codec,
                    sampleRate: stream.audio.sample_rate,
                    channels: stream.audio.channel,
                } : undefined
            }));
        } catch (error) {
            throw new Error(`Failed to get SRS streams info: ${error}`);
        }
    }

    async getClientInfo(): Promise<ClientInfo[]> {
        try {
            const response = await axios.get<SRSClientsResponse>(`${this.baseUrl}/api/v1/clients/`);
            return response.data.clients
                .filter(client => !client.publish)
                .map(client => {
                    console.log('Client alive value:', client.alive, typeof client.alive);
                    return {
                        id: client.id,
                        vhost: client.vhost,
                        stream: client.stream,
                        ip: client.ip,
                        url: client.url,
                        alive: Math.round(client.alive * 1000), // 转换为毫秒并四舍五入
                        type: client.type
                    };
                });
        } catch (error) {
            throw new Error(`Failed to get SRS clients info: ${error}`);
        }
    }
} 